# BIG DATA ANALYSIS

# POLICE DATASET ANALYSIS

here
 
 This data is available as a CSV file. we are going to analysis this data  through pandas DataFrame.

In [1]:
     import pandas as pd

In [2]:
df=pd.read_csv('E:\\kaggle dataset\\Police Data.csv')

### Data Cleaning

# Analysis the data set of Police

In [3]:
df.head(2)

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0   1/2/2005      1:55           NaN             M          1985.0   
1  1/18/2005      8:15           NaN             M          1965.0   

   driver_age driver_race violation_raw violation  search_conducted  \
0        20.0       White      Speeding  Speeding             False   
1        40.0       White      Speeding  Speeding             False   

  search_type stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN     Citation       False      0-15 Min               False  
1         NaN     Citation       False      0-15 Min               False

In [4]:
df.shape

(65535, 15)

# By Index Attribute provide the index of dataset

In [5]:
df.index

RangeIndex(start=0, stop=65535, step=1)

'dtypes' it show data type of each coloum

In [6]:
df.dtypes

stop_date              object
stop_time              object
country_name          float64
driver_gender          object
driver_age_raw        float64
driver_age            float64
driver_race            object
violation_raw          object
violation              object
search_conducted         bool
search_type            object
stop_outcome           object
is_arrested            object
stop_duration          object
drugs_related_stop       bool
dtype: object

In [7]:
df.nunique()

stop_date             2651
stop_time             1432
country_name             0
driver_gender            2
driver_age_raw          93
driver_age              73
driver_race              5
violation_raw           12
violation                6
search_conducted         2
search_type             23
stop_outcome             6
is_arrested              2
stop_duration            4
drugs_related_stop       2
dtype: int64

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65535 entries, 0 to 65534
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   stop_date           65535 non-null  object 
 1   stop_time           65535 non-null  object 
 2   country_name        0 non-null      float64
 3   driver_gender       61474 non-null  object 
 4   driver_age_raw      61481 non-null  float64
 5   driver_age          61228 non-null  float64
 6   driver_race         61475 non-null  object 
 7   violation_raw       61475 non-null  object 
 8   violation           61475 non-null  object 
 9   search_conducted    65535 non-null  bool   
 10  search_type         2479 non-null   object 
 11  stop_outcome        61475 non-null  object 
 12  is_arrested         61475 non-null  object 
 13  stop_duration       61475 non-null  object 
 14  drugs_related_stop  65535 non-null  bool   
dtypes: bool(2), float64(3), object(10)
memory usage: 6.6+

# Delete the Coloumn in which have null value

In [9]:
df.head(2)

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0   1/2/2005      1:55           NaN             M          1985.0   
1  1/18/2005      8:15           NaN             M          1965.0   

   driver_age driver_race violation_raw violation  search_conducted  \
0        20.0       White      Speeding  Speeding             False   
1        40.0       White      Speeding  Speeding             False   

  search_type stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN     Citation       False      0-15 Min               False  
1         NaN     Citation       False      0-15 Min               False

In [10]:
df.isnull().sum()

stop_date                 0
stop_time                 0
country_name          65535
driver_gender          4061
driver_age_raw         4054
driver_age             4307
driver_race            4060
violation_raw          4060
violation              4060
search_conducted          0
search_type           63056
stop_outcome           4060
is_arrested            4060
stop_duration          4060
drugs_related_stop        0
dtype: int64

In [11]:
# df.drop(columns='country_name',inplace=True)
df.drop(columns='search_type',inplace=True)

### After delete the 2 columns of data 

In [12]:
df.head(3)

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0   1/2/2005      1:55           NaN             M          1985.0   
1  1/18/2005      8:15           NaN             M          1965.0   
2  1/23/2005     23:15           NaN             M          1972.0   

   driver_age driver_race violation_raw violation  search_conducted  \
0        20.0       White      Speeding  Speeding             False   
1        40.0       White      Speeding  Speeding             False   
2        33.0       White      Speeding  Speeding             False   

  stop_outcome is_arrested stop_duration  drugs_related_stop  
0     Citation       False      0-15 Min               False  
1     Citation       False      0-15 Min               False  
2     Citation       False      0-15 Min               False

# for speeding where the man and woman stopped more often

In [13]:
df.head(2)

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0   1/2/2005      1:55           NaN             M          1985.0   
1  1/18/2005      8:15           NaN             M          1965.0   

   driver_age driver_race violation_raw violation  search_conducted  \
0        20.0       White      Speeding  Speeding             False   
1        40.0       White      Speeding  Speeding             False   

  stop_outcome is_arrested stop_duration  drugs_related_stop  
0     Citation       False      0-15 Min               False  
1     Citation       False      0-15 Min               False

In [14]:
df[df.violation == 'Speeding'].driver_gender.value_counts()

M    25517
F    11686
Name: driver_gender, dtype: int64

### man have a more often than the female male have a 25517 and woman have 11686

# Does gender affect who gets searched during a stop

In [15]:
df.head(2)

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0   1/2/2005      1:55           NaN             M          1985.0   
1  1/18/2005      8:15           NaN             M          1965.0   

   driver_age driver_race violation_raw violation  search_conducted  \
0        20.0       White      Speeding  Speeding             False   
1        40.0       White      Speeding  Speeding             False   

  stop_outcome is_arrested stop_duration  drugs_related_stop  
0     Citation       False      0-15 Min               False  
1     Citation       False      0-15 Min               False

In [16]:
df.search_conducted.value_counts()

False    63056
True      2479
Name: search_conducted, dtype: int64

In [17]:
df.groupby('driver_gender').search_conducted.value_counts()

driver_gender  search_conducted
F              False               15944
               True                  366
M              False               43051
               True                 2113
Name: search_conducted, dtype: int64

### here we can check the Male search conducted is 2213 and Female is 366 only 

# what is mean stop duration.

df.head(2)

#### stop duration is in String so that first we convert it in the int

In [18]:
 df.stop_duration.value_counts()

0-15 Min     47379
16-30 Min    11448
30+ Min       2647
2                1
Name: stop_duration, dtype: int64

####  there have 4 uniques value pesent in the stop_duration column

In [19]:
df['stop_duration']=df['stop_duration'].map({'0-15 Min':7.5 ,'16-30 Min':23 ,'30+ Min':45,'2':2 })

In [20]:
df

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0       1/2/2005      1:55           NaN             M          1985.0   
1      1/18/2005      8:15           NaN             M          1965.0   
2      1/23/2005     23:15           NaN             M          1972.0   
3      2/20/2005     17:15           NaN             M          1986.0   
4      3/14/2005     10:00           NaN             F          1984.0   
...          ...       ...           ...           ...             ...   
65530  12/6/2012     17:54           NaN             F          1987.0   
65531  12/6/2012     22:22           NaN             M          1954.0   
65532  12/6/2012     23:20           NaN             M          1985.0   
65533  12/7/2012      0:23           NaN           NaN             NaN   
65534  12/7/2012      0:30           NaN             F          1985.0   

       driver_age driver_race                   violation_raw  violation  \
0            20.0       White                        Speeding   Speeding   
1            40.0       White                        Speeding   Speeding   
2            33.0       White                        Speeding   Speeding   
3            19.0       White                Call for Service      Other   
4            21.0       White                        Speeding   Speeding   
...           ...         ...                             ...        ...   
65530        25.0       White                        Speeding   Speeding   
65531        58.0       White                        Speeding   Speeding   
65532        27.0       Black  Equipment/Inspection Violation  Equipment   
65533         NaN         NaN                             NaN        NaN   
65534        27.0       White                        Speeding   Speeding   

       search_conducted   stop_outcome is_arrested  stop_duration  \
0                 False       Citation       False            7.5   
1                 False       Citation       False            7.5   
2                 False       Citation       False            7.5   
3                 False  Arrest Driver        True           23.0   
4                 False       Citation       False            7.5   
...                 ...            ...         ...            ...   
65530             False       Citation       False            7.5   
65531             False        Warning       False            7.5   
65532             False       Citation       False            7.5   
65533             False            NaN         NaN            NaN   
65534             False       Citation       False            7.5   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65530               False  
65531               False  
65532               False  
65533               False  
65534               False  

[65535 rows x 14 columns]

In [21]:
df['stop_duration'].value_counts()

7.5     47379
23.0    11448
45.0     2647
2.0         1
Name: stop_duration, dtype: int64

#### the string values change in int so now we can find the mean of stop_duration

In [22]:
df['stop_duration'].mean()

12.001032940219602

 # Compare the Age ditribution on each violation

In [23]:
df.head(2)

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0   1/2/2005      1:55           NaN             M          1985.0   
1  1/18/2005      8:15           NaN             M          1965.0   

   driver_age driver_race violation_raw violation  search_conducted  \
0        20.0       White      Speeding  Speeding             False   
1        40.0       White      Speeding  Speeding             False   

  stop_outcome is_arrested  stop_duration  drugs_related_stop  
0     Citation       False            7.5               False  
1     Citation       False            7.5               False

In [24]:
df.groupby('driver_age').violation.value_counts()

driver_age  violation          
15.0        Moving violation        4
            Speeding                1
16.0        Speeding               18
            Moving violation        6
            Registration/plates     5
                                   ..
85.0        Moving violation        1
86.0        Speeding                3
            Moving violation        2
            Other                   1
88.0        Speeding                2
Name: violation, Length: 333, dtype: int64

In [25]:
df.groupby('violation').driver_age.describe()

count       mean        std   min   25%   50%   75%  \
violation                                                                    
Equipment             6507.0  31.682957  11.380671  16.0  23.0  28.0  39.0   
Moving violation     11876.0  36.736443  13.258350  15.0  25.0  35.0  47.0   
Other                 3477.0  40.362381  12.754423  16.0  30.0  41.0  50.0   
Registration/plates   2240.0  32.656696  11.150780  16.0  24.0  30.0  40.0   
Seat belt                3.0  30.333333  10.214369  23.0  24.5  26.0  34.0   
Speeding             37120.0  33.262581  12.615781  15.0  23.0  30.0  42.0   

                      max  
violation                  
Equipment            81.0  
Moving violation     86.0  
Other                86.0  
Registration/plates  74.0  
Seat belt            42.0  
Speeding             88.0